File: frontend/lib/db/paths.mjs
import path from 'node:path'
import fs from 'node:fs'
import { fileURLToPath } from 'node:url'

const here = path.dirname(fileURLToPath(import.meta.url))

/** Repo root — `frontend/lib/db` → up three levels. */
export const repoRoot = path.resolve(here, '..', '..', '..')

/**
 * Absolute path of the SQLite file. Override with DATABASE_PATH (relative paths
 * resolve against the repo root) — useful when the repo lives in a synced folder
 * like OneDrive, where SQLite's file locking is unreliable.
 */
export function resolveDbPath() {
  const configured = process.env.DATABASE_PATH
  const target = configured
    ? path.isAbsolute(configured)
      ? configured
      : path.resolve(repoRoot, configured)
    : path.join(repoRoot, 'data', 'daily-trends.db')
  fs.mkdirSync(path.dirname(target), { recursive: true })
  return target
}

export const snapshotPath = path.join(repoRoot, 'backend', 'scraper', 'latest-products.json')
export const taxonomyPath = path.join(repoRoot, 'shared', 'taxonomy.json')
export const curatedAdsPath = path.join(repoRoot, 'shared', 'curated-ads.json')
export const mailOutboxDir = path.join(repoRoot, 'data', 'mail-outbox')

/** Reads a JSON file from the shared/ directory at runtime. */
export function readSharedJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf8'))
}

File: backend/scraper/paths.py
"""Shared filesystem locations. Mirrors frontend/lib/db/paths.mjs."""

import os

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(BASE_DIR, "..", ".."))

SNAPSHOT_PATH = os.path.join(BASE_DIR, "latest-products.json")
TAXONOMY_PATH = os.path.join(REPO_ROOT, "shared", "taxonomy.json")
CURATED_ADS_PATH = os.path.join(REPO_ROOT, "shared", "curated-ads.json")


def resolve_db_path() -> str:
    """Absolute path of the SQLite file.

    Honours DATABASE_PATH exactly like the Node side does (relative paths
    resolve against the repo root), so both processes always open the same file.
    """
    configured = os.getenv("DATABASE_PATH")
    if configured:
        target = configured if os.path.isabs(configured) else os.path.join(REPO_ROOT, configured)
    else:
        target = os.path.join(REPO_ROOT, "data", "daily-trends.db")

    os.makedirs(os.path.dirname(target), exist_ok=True)
    return target
